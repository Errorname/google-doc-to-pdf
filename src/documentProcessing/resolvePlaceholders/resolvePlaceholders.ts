import { Placeholder, Formatters } from '../types'

import computeContent from './computeContent'

const resolvePlaceholders = (
  placeholders: Placeholder[],
  data: Record<string, any>,
  options?: {
    formatters?: Formatters
    strict?: boolean
  }
): Placeholder[] => {
  return placeholders.map(placeholder => {
    const rawWithoutBrackets = placeholder.raw.slice(2, -2).trim()

    // CommentPlaceholder
    if (rawWithoutBrackets.startsWith('#')) {
      return {
        ...placeholder,
        type: 'comment',
        value: rawWithoutBrackets.slice(1).trim()
      }
    }

    if (['%', '@'].includes(rawWithoutBrackets[0])) {
      // Reserved for later usage
      return placeholder
    }

    // ContentPlaceholder
    return computeContent(placeholder, data, options)
  })
}

export default resolvePlaceholders
