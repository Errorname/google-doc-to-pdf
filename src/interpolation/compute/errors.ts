export class UnknownFormatterError extends Error {}

export class UnvalidFormatterError extends Error {}

export class UndefinedVariableError extends Error {}
