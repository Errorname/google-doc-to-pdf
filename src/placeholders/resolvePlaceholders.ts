import { Formatters, Placeholder, TypedPlaceholder } from '../types'
import { compute, parser } from './syntax'

const resolvePlaceholders = (
  placeholders: Placeholder[],
  data: Record<string, unknown>,
  options: {
    formatters: Formatters
    strict: boolean
  }
): TypedPlaceholder[] => {
  return placeholders.map((placeholder) => {
    const rawWithoutBrackets = placeholder.raw.slice(2, -2).trim()

    // Comment Placeholder
    if (rawWithoutBrackets.startsWith('#')) {
      return {
        ...placeholder,
        type: 'comment',
        value: rawWithoutBrackets.slice(1).trim(),
      }
    }

    // Future Placeholders
    if (['%', '@'].includes(rawWithoutBrackets[0])) {
      // Reserved for later usage
      return {
        ...placeholder,
        type: 'unknown',
      }
    }

    // Content Placeholder
    const expression = parser(rawWithoutBrackets)
    const result = compute(expression, data, options)
    return {
      ...placeholder,
      type: 'content',
      value: result,
    }
  })
}

export default resolvePlaceholders
