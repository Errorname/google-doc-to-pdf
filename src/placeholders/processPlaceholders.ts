import { Formatters, GDoc, TypedPlaceholder } from '../types'
import findPlaceholders from './findPlaceholders'
import resolvePlaceholders from './resolvePlaceholders'

const processPlaceholders = (
  doc: GDoc,
  data: Record<string, unknown>,
  formatters: Formatters,
  strict: boolean
): TypedPlaceholder[] => {
  const placeholders = findPlaceholders(doc)
  return resolvePlaceholders(placeholders, data, { formatters, strict })
}

export default processPlaceholders
