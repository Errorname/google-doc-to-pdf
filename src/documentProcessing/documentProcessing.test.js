import processPlaceholders, { findPlaceholders, buildUpdates } from './documentProcessing'
import { UndefinedVariableError, UnknownFormatterError } from './resolvePlaceholders/errors'

describe('documentProcessing', () => {
  test('full document', () => {
    const document = {
      body: {
        content: [
          {
            paragraph: {
              elements: [
                { textRun: { content: 'Hello {{ user.name }}!' }, startIndex: 0, endIndex: 21 },
                {}
              ]
            }
          },
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: 'Have you seen {{movie.title | uppercase}}?' },
                  startIndex: 23,
                  endIndex: 64
                },
                {
                  textRun: {
                    content: 'In my opinion, I would rate it {{ movie.rating }} out of 5',
                    startIndex: 66,
                    endIndex: 123
                  }
                }
              ]
            }
          },
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: '{{ movie.poster | image }}' },
                  startIndex: 124,
                  endIndex: 150
                }
              ]
            }
          }
        ]
      }
    }
    const result = processPlaceholders(document, {
      user: { name: 'Thibaud' },
      movie: { title: 'Lost in Translation', rating: 5, poster: 'http://example.img' }
    })
    const updates = buildUpdates(result)
    expect(result).toMatchSnapshot()
    expect(updates).toMatchSnapshot()
  })

  test('full document — strict - Unknown formatter', () => {
    const document = {
      body: {
        content: [
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: 'Hello {{ user.name | pretty }}!' },
                  startIndex: 0,
                  endIndex: 21
                },
                {}
              ]
            }
          },
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: 'Have you seen {{movie.title | uppercase}}?' },
                  startIndex: 23,
                  endIndex: 64
                },
                {
                  textRun: {
                    content: 'In my opinion, I would rate it {{ movie.rating }} out of 5',
                    startIndex: 66,
                    endIndex: 123
                  }
                }
              ]
            }
          },
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: '{{ movie.poster | image }}' },
                  startIndex: 124,
                  endIndex: 150
                }
              ]
            }
          }
        ]
      }
    }
    expect(() => {
      const result = processPlaceholders(
        document,
        {
          user: { name: 'Thibaud' },
          movie: { title: 'Lost in Translation', rating: 5, poster: 'http://example.img' }
        },
        {},
        true
      )
    }).toThrow(UnknownFormatterError)
  })

  test('full document — strict - Undefined variable', () => {
    const document = {
      body: {
        content: [
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: 'Hello {{ user.surname }}!' },
                  startIndex: 0,
                  endIndex: 21
                },
                {}
              ]
            }
          },
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: 'Have you seen {{ movie.title | uppercase }}?' },
                  startIndex: 23,
                  endIndex: 64
                },
                {
                  textRun: {
                    content: 'In my opinion, I would rate it {{ movie.rating }} out of 5',
                    startIndex: 66,
                    endIndex: 123
                  }
                }
              ]
            }
          },
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: '{{ movie.poster | image }}' },
                  startIndex: 124,
                  endIndex: 150
                }
              ]
            }
          }
        ]
      }
    }
    expect(() => {
      const result = processPlaceholders(
        document,
        {
          user: { name: 'Thibaud' },
          movie: { title: 'Lost in Translation', rating: 5, poster: 'http://example.img' }
        },
        {},
        true
      )
    }).toThrow(UndefinedVariableError)
  })

  test('full document — strict - First error is thrown first', () => {
    const document = {
      body: {
        content: [
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: 'Hello {{ user.surname }}!' },
                  startIndex: 0,
                  endIndex: 21
                },
                {}
              ]
            }
          },
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: 'Have you seen {{ movie.title | pretty }}?' },
                  startIndex: 23,
                  endIndex: 64
                },
                {
                  textRun: {
                    content: 'In my opinion, I would rate it {{ movie.rating }} out of 5',
                    startIndex: 66,
                    endIndex: 123
                  }
                }
              ]
            }
          },
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: '{{ movie.poster | image }}' },
                  startIndex: 124,
                  endIndex: 150
                }
              ]
            }
          }
        ]
      }
    }
    expect(() => {
      const result = processPlaceholders(
        document,
        {
          user: { name: 'Thibaud' },
          movie: { title: 'Lost in Translation', rating: 5, poster: 'http://example.img' }
        },
        {},
        true
      )
    }).toThrow(UndefinedVariableError)
  })
})
