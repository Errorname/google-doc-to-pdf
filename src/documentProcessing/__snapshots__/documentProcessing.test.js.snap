// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`documentProcessing full document 1`] = `
Array [
  Object {
    "input": Object {
      "raw": "user.name",
      "value": "Thibaud",
    },
    "output": "Thibaud",
    "pipes": Array [],
    "position": Object {
      "end": 21,
      "start": 6,
    },
    "raw": "{{ user.name }}",
    "type": "content",
  },
  Object {
    "input": Object {
      "raw": "movie.title",
      "value": "Lost in Translation",
    },
    "output": "LOST IN TRANSLATION",
    "pipes": Array [
      Object {
        "output": "LOST IN TRANSLATION",
        "raw": "uppercase",
      },
    ],
    "position": Object {
      "end": 64,
      "start": 37,
    },
    "raw": "{{movie.title | uppercase}}",
    "type": "content",
  },
  Object {
    "input": Object {
      "raw": "movie.rating",
      "value": 5,
    },
    "output": 5,
    "pipes": Array [],
    "position": Object {
      "end": NaN,
      "start": NaN,
    },
    "raw": "{{ movie.rating }}",
    "type": "content",
  },
  Object {
    "input": Object {
      "raw": "movie.poster",
      "value": "http://example.img",
    },
    "output": Object {
      "height": undefined,
      "url": "http://example.img",
      "width": undefined,
    },
    "pipes": Array [
      Object {
        "output": Object {
          "height": undefined,
          "url": "http://example.img",
          "width": undefined,
        },
        "raw": "image",
      },
    ],
    "position": Object {
      "end": 150,
      "start": 124,
    },
    "raw": "{{ movie.poster | image }}",
    "type": "content",
  },
]
`;

exports[`documentProcessing full document 2`] = `
Array [
  Object {
    "insertInlineImage": Object {
      "location": Object {
        "index": 124,
        "segmentId": "",
      },
      "objectSize": Object {
        "height": undefined,
        "width": undefined,
      },
      "uri": "http://example.img",
    },
  },
  Object {
    "replaceAllText": Object {
      "containsText": Object {
        "matchCase": true,
        "text": "{{ user.name }}",
      },
      "replaceText": "Thibaud",
    },
  },
  Object {
    "replaceAllText": Object {
      "containsText": Object {
        "matchCase": true,
        "text": "{{movie.title | uppercase}}",
      },
      "replaceText": "LOST IN TRANSLATION",
    },
  },
  Object {
    "replaceAllText": Object {
      "containsText": Object {
        "matchCase": true,
        "text": "{{ movie.rating }}",
      },
      "replaceText": "5",
    },
  },
  Object {
    "replaceAllText": Object {
      "containsText": Object {
        "matchCase": true,
        "text": "{{ movie.poster | image }}",
      },
      "replaceText": "",
    },
  },
]
`;
